agent/task.go

```diff
@@ -201,6 +201,17 @@ func (tm *taskManager) run(ctx context.Context) {
 			}
 		case <-shutdown:
 			if cancel != nil {
+				// Wait for operation to finish if the task
+				// is in preparing state. This avoid the race
+				// between Prepare and Remove operations,
+				// which could cause orphan containers.
+				if tm.task.Status.State == api.TaskStatePreparing {
```

Comment:
I agree it's better doing it in exec.Do so that it can go through the state machine. I thought about it but I didn't figure out how. It looks the shutdown is handled as an special event which can abort other ongoing activities for the same task.